scripts/aot_albert.py

```diff
@@ -9,79 +9,106 @@
-
-clear_params(aot_model)
-torch.manual_seed(0)
-out2 = aot_model(**train_inputs).loss
-out2.sum().backward()
-grad2 = [i.grad for i in aot_model.parameters()]
-print((out1 - out2).abs().max())
-print(max([(a-b).abs().max() for a, b in zip(grad1, grad2)]))
+        train_inputs = {'input_ids': input_ids, 'labels': decoder_ids}
+
+        def get_cur_memory():
+            torch.cuda.synchronize()
+            import gc; gc.collect()
+            torch.cuda.empty_cache()
+            stats = torch.cuda.memory_stats()
+            peak_bytes_requirement = stats["allocated_bytes.all.current"]
+            print(f"Current memory requirement: {peak_bytes_requirement / 1024 ** 3:.2f} GB")
+            return peak_bytes_requirement
+
+        def bench_model(name, mod):
+            m = None
+            for i in range(5):
+                out = mod(**train_inputs).loss.abs().sum()
+                if i == 4:
+                    m = get_cur_memory()
+                out.backward()
+            iters = 20
+            torch.cuda.synchronize()
+            begin = time.time()
+            for _ in range(iters):
+                mod(**train_inputs).loss.sum().backward()
+            torch.cuda.synchronize()
+            t = (time.time()-begin)/iters
+            print(name, (time.time()-begin)/iters)
+            return t, m
+
+        aot_model = memory_efficient_fusion(model)
+        model_name = type(model).__name__
+        t,m = bench_model("eager", model)
+        model_header, dtype_header, nh, th, mh, tp, mp = "model", "dtype", "name", "time (s)", "mem (GB)", "time %", "mem %"
+        results.append({model_header: model_name, dtype_header: str(dtype), nh:"eager", th:t, mh:m/2**30})
+        with torch.jit.fuser("fuser2"):
+            t,m = bench_model("aot", aot_model)
+        results.append({model_header: model_name, dtype_header: str(dtype), nh:"aot", th:t, mh:m/2**30})
+
+        # calculate relative improvements
+        base_r = results[-2]
+        for r in results[-2:]:
+            r[tp] = round(100 * (r[th] - base_r[th]) / base_r[th])
+            r[mp] = round(100 * (r[mh] - base_r[mh]) / base_r[mh])
+        print(pd.DataFrame(results[-2:]).to_markdown(index=False, floatfmt=".3f"))
+
+        print()
+
+        ## Checking correctness
+        def clear_params(model):
+            for i in model.parameters(): i.grad = None
+
+        clear_params(model)
+        torch.manual_seed(0)
+        out1 = model(**train_inputs).loss
+        out1.sum().backward()
+        grad1 = [i.grad for i in model.parameters()]
+
+        clear_params(aot_model)
+        torch.manual_seed(0)
+        out2 = aot_model(**train_inputs).loss
+        out2.sum().backward()
+        grad2 = [i.grad for i in aot_model.parameters()]
+        print("Maximum output error: ", (out1 - out2).abs().max().item())
+        print("Maximum gradient error: ", max([(a-b).abs().max() for a, b in zip(grad1, grad2) if a is not None]).item())
```

Comment:
may I suggest to actually set the expected tolerance and do `torch.testing.assert_close` check, since it's hard to notice visually when something is wrong, especially now that we have a lot more of these. and then it just becomes noise.

Reply:
Indeed. Just been there and `torch.testing.assert_close` is pretty handy.

Reply:
Hmm, any thoughts on rtol/atol values? It's kind of difficult to test these kinds of gradients on whole models, since catastrophic cancellation can lead to relatively large atol values and very large rtol values.

Reply:
Updated with some tolerance checks, but they're much looser than I would prefer lol.

Reply:
I was hoping you had a much better sense of what the ideal values would be. Let's start with your proposed setup and then adapt as needed.

The next level of testing is to run some scoring function like glue on a real data and compare the output is the same or is very similar. but in other ways it can be less precise as well. let's keep on experimenting.